### PR TITLE
Check that `SingleThreadedBlasScope` is created and destroyed on the same `std::thread`

### PR DESCRIPTION
**Pull request by msimberg**
Based on https://github.com/eth-cscs/DLA-Future/pull/860#discussion_r1209860337.

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

File: src/common/single_threaded_blas.cpp [hunk 1]
```diff
@@ -15,10 +15,14 @@
 #include <mkl_service.h>
 #endif
 
+#include <dlaf/common/assert.h>
```
Comment (RMeli) [inline src/common/single_threaded_blas.cpp:18]: Does this need `#ifdef DLAF_ASSERT_MODERATE_ENABLE`?

**Review by RMeli — APPROVED** [review]

File: src/common/single_threaded_blas.cpp (hunk 1 above)
Comment (msimberg) [reply src/common/single_threaded_blas.cpp]: Good point. Not necessary, but it doesn't hurt. I added the `ifdef`.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
@msimberg FYI:
```
test 43
      Start 43: test_bt_band_to_tridiag
43: Test command: /DLA-Future-build/test/unit/eigensolver/test_bt_band_to_tridiag
43: Test timeout computed to be: 10000000
43: Running main() from gtest_mpipika_main.cpp
43: [==========] Running 16 tests from 4 test suites.
43: [----------] Global test environment set-up.
43: [----------] 4 tests from BacktransformationBandToTridiagTestMC/0, where TypeParam = float
43: [ RUN      ] BacktransformationBandToTridiagTestMC/0.CorrectnessLocal
43: [ERROR] /DLA-Future/src/common/single_threaded_blas.cpp:41 : dlaf::common::internal::SingleThreadedBlasScope::~SingleThreadedBlasScope()
43: calling_thread == current_thread
43: 140727143212608
43: 140727134819904
43: 
43: terminate called without an active exception
13/27 Test #43: test_bt_band_to_tridiag ...........***Failed    0.38 sec
```

**Comment by msimberg** [comment]
cscs-ci rnu

**Comment by msimberg** [comment]
Yielding was happening in the test itself of `bt_band_to_tridiag` (not the algorithm implementation) and I've fixed that in https://github.com/eth-cscs/DLA-Future/pull/889/commits/6daae56f823139ea5ba23c328dd505d502e562b6.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
I had to fix one more yielding while holding `SingleThreadedBlasScope` in `test_tridiag_solver_rot`. Note that the test was hanging rather than directly reporting the failure and I'm not quite sure why that's the case. The reason for the failure was the same though. Locally it failed immediately whereas in CI it hung.

**Review by albestro — APPROVED** [review]

**Review by rasolca — APPROVED** [review]

**Pull request merged**